File: 2024/day03/main.lua
local function prod(mem)
  local sum = 0
  for x, y in string.gmatch(mem, 'mul%((%d+),(%d+)%)') do
    sum = sum + tonumber(x) * tonumber(y)
  end
  return sum
end

local file = io.open('input', 'r')
local mem = file:read('*a')
file:close()

-- Part 1: 173517243
print(prod(mem))

-- Part 2: 100450138
local sum = 0
while true do
  local start = string.find(mem, "don't%(%)")
  if not start then
    break
  end
  sum = sum + prod(mem:sub(1, start - 1))
  local end_pos = string.find(mem, 'do%(%)', start)
  if not end_pos then
    break
  end
  mem = mem:sub(end_pos + 3)
end
print(sum)

File: 2024/day03/main.py
from re import compile

with open('input', 'r') as file:
	mem = file.read()


def prod(mem):
	sum = 0
	matches = compile(r'mul\(\d+,\d+\)').findall(mem)
	for match in matches:
		x, y = map(int, match[4:-1].split(','))
		sum += x * y
	return sum


# Part 1: 173517243
print(prod(mem))

# Part 2: 100450138
sum = 0
while True:
	start = mem.find("don't()")
	if start == -1:
		break
	sum += prod(mem[:start])
	end = mem.find('do()', start)
	if end == -1:
		break
	mem = mem[end + 4 :]
print(sum)
